### ISSUE TITLE
Support multi-file validation Checks in Checkstyle

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#3540]
Right now Checkstyle is single-file validation tool, with cache that heavily depends on this design pattern.

We need to find solution to support multi-file validation Checks.

**Comment by GitToasterhub** [comment]
Oh.This issue is quite interesting and challenging.What is interesting for me,can we consider this issue like an idea for GSOC?

**Comment by romani** [comment]
@GitToasterhub , please stay off issues that do not have "approved" label.
This issue is very complicated , there is no even good idea on how to make it from admins of project. 

GSoC have only doable and clear tasks that has to be done and merged to repo on the moment of GSoC completion. List of GSoC projects will not be changed.

**Comment by romani** [comment]
additional problem is multi file support is that when all violations are collected from file and put them to loggers. in case of XML logger, it opens and close `file` tag. So reporting violation from multi-file Checks might also to the same files so in XML we will have multiple tags for the same file. Order of violations in Default logger might also be unsorted, there will be grouped by file violations and then random order of violations from multi-file Checks.

**Comment by rnveach** [comment]
> reporting violation from multi-file Checks might also to the same files so in XML we will have multiple tags for the same file.
> they will be grouped by file violations and then random order of violations from multi-file Checks.

This can be seen for TranslationCheck as it is already doing this.
I have no problems with this random violation order in my own project. All violations need to be corrected in the end anyways.

**Comment by romani** [comment]
After long thoughts ..... I am ok to extend checkstyle to be multi-file validator, as for now only text processing (no class-path).

some details of current hacks are at https://github.com/checkstyle/checkstyle/issues/4869#issuecomment-396826293
all "protected" method should be removed.
We change API of `void finishProcessing()` to be `SortedSet<LocalizedMessage> finishProcessing()`.

@rnveach , if you are agree, please feel free to make issue as approved.

**Comment by rnveach** [comment]
I agree, but https://github.com/checkstyle/checkstyle/issues/3539 also needs to be solved for full multi-file support, otherwise the cache will either hide violations or create false violations.

> all "protected" method should be removed.
We change API of void finishProcessing() to be SortedSet<LocalizedMessage> finishProcessing().

As a multi-file module, it is the job the FileSet/Check to call `fireFileStarted` and `fireFileFinished` as only it knows what files it is processing again. Without  `AbstractFileSetCheck.fireErrors`, the current `log` implementation doesn't know which file it is working with. To remove it, `LocalizedMessage` would have to include the file name, and Checker would have to do the fire calls itself and group the messages by file name.

**Comment by rnveach** [comment]
No that is everything.

> Lets change it on this level first and see what other conner case we forgot

This is similar to how I have implemented and there hasn't been any issues. `finishProcessing` is what all modules need to implement multi-file support.

**Comment by romani** [comment]
> finishProcessing is what all modules need to implement multi-file support.

for FileSet Check yes.
For Java checks ... I would rather postpone this support a bit and collect user storied on what validations they would like to make.

---------------

if you agree, please label issue as approved.

**Comment by rnveach** [comment]
> for FileSet Check yes. For Java checks ... 

I was referring to both.

**Comment by romani** [comment]
This issue is blocked by cache usage discussion at #3539.
As we agree on cache role for single-file Checks and multi-file Checks, we can proceed with this issue.

**Comment by rnveach** [comment]
Adding support will need to handle exceptions properly as seen in https://github.com/checkstyle/checkstyle/issues/6340#issuecomment-450937749 . Checker should handle exceptions itself like it does for non-multi-file checks and not require the checks to change them into violations. Halt on exception should be looked into for this too.
